### PR TITLE
Check the right kind of error is thrown in tests

### PR DESCRIPTION
**Pull request by hdgarrood**
Add a basic mechanism for declaring what kind of errors a test should
throw, and add a few of these declarations to failing examples.

If you think this looks good I'll go ahead and add these declarations to all of the failing examples.

Refs (possibly fixes?) #1181

### CODE REVIEW

**Comment by paf31** [comment]
Looks fantastic, thank you! Once we have all tests marked up, perhaps it makes sense to use the lack of an annotation to mean "should pass"?

**Comment by michaelficarra** [comment]
@hdgarrood Can you show some examples of multiple errors or compilation errors?

**Comment by hdgarrood** [comment]
> Once we have all tests marked up, perhaps it makes sense to use the lack of an annotation to mean "should pass"?

I'm not sure, we do already have a way of indicating this based on what directory the file is in; that is, `passing` vs `failing`. I do think it would be a good idea to make sure that every failing test does fail with a specific error though; perhaps a good way to achieve this would be to require that every test inside the `failing` directory has at least one shouldFailWith declaration?

> Can you show some examples of multiple errors or compilation errors?

I've put 3 examples in already, I'm going to start going through and adding more now. Come to think of it I'm not sure if there are any that should produce more than 1 error, we'll see in a moment.

File: examples/failing/MultipleErrors.purs [hunk 1]
```diff
@@ -1,11 +1,13 @@
+-- @shouldFailWith ExprDoesNotHaveType
```
Comment (hdgarrood) [inline examples/failing/MultipleErrors.purs]: @michaelficarra Here's an example with a test raising multiple errors.

**Comment by hdgarrood** [comment]
Quite a few of these no longer testing what they are supposed to test since 0.7! I think this is all good now, though.

**Comment by hdgarrood** [comment]
Also, some next steps from here:
* add tests for suggestions (I think we should have a real data type for suggestions too, not just strings).
* add tests for warnings
* work out how to run all the tests in the manual directory automatically. Perhaps this mechanism is enough, but we can presumably also embed specific examples inside the Haskell source if we need a more elaborate check.

**Comment by paf31** [comment]
Brilliant! I've wanted this for so long and never got round to implementing it. Thanks!

**Pull request merged**